Errors aren't detected if media element isn't in the DOM.

jyavenard opened the pull request:
#### 20984f625b7f1108a857f3c023e9c1cad37ec6be
<pre>
Errors aren&apos;t detected if media element isn&apos;t in the DOM.
<a href="https://bugs.webkit.org/show_bug.cgi?id=283150">https://bugs.webkit.org/show_bug.cgi?id=283150</a>
<a href="https://rdar.apple.com/139935149">rdar://139935149</a>

Reviewed by Eric Carlson.

When the video element isn&apos;t in the DOM, we use a WebCoreDecompressionSession
to decode the content in place of an AVSampleBufferDisplayLayer.
In the common case with an AVSBDL, we listen to layer errors published by the NotificationCentre.
When a WebCoreDecompressionSession is in use, no such events are fired.
The WebCoreDecompressionSession provides a listener for errors, but it was unused functionality.

We add an error handler to the WebCoreDecompressionSession used by their respective MediaPlayerPrivate
which will allow the error to be handled and reported to the HTMLMediaElement.

Added tests.

* LayoutTests/media/media-source/media-source-vp8-webm-error-expected.txt: Added.
* LayoutTests/media/media-source/media-source-vp8-webm-error-offscreen-expected.txt: Added.
* LayoutTests/media/media-source/media-source-vp8-webm-error-offscreen.html: Added.
* LayoutTests/media/media-source/media-source-vp8-webm-error.html: Added.
* LayoutTests/media/media-vp8-webm-error-offscreen-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-error-offscreen.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::ensureDecompressionSession):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::ensureDecompressionSession):

Canonical link: <a href="https://commits.webkit.org/286685@main">https://commits.webkit.org/286685@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71045b500bee969e834ecacf14d2f4991ca9862b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55632 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29503 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81127 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27873 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78714 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64774 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3925 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60049 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18150 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79664 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49966 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65766 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40374 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47366 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23259 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26197 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68488 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82573 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3973 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2617 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68326 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4126 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65738 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67575 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16879 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11546 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9631 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3920 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6729 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3943 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7373 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5701 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1535,6 +1535,13 @@ static bool isCopyDisplayedPixelBufferAvailable()
+    m_decompressionSession->setErrorListener([weakThis = WeakPtr { *this }](OSStatus) {
+        if (RefPtr protectedThis = weakThis.get()) {
+            protectedThis->setNetworkState(MediaPlayer::NetworkState::DecodeError);
+            protectedThis->setReadyState(MediaPlayer::ReadyState::HaveNothing);
```
Why not setReadyState in MediaPlayerPrivateMediaSourceAVFObjC.mm?
Also, can you beef up the commit log, it is not clear to me why the error listener is only needed for media elements that are not in the DOM.

jyavenard replied:
Because when it's not in the DOM we use a WebCoreDecompressionSession if we have an AVSBDL, the error are handled by the SourceBufferPrivateAVFObjC listener to the NotificationCentre for errors.

As for why the readyState isn't changed for MSE, I made it consistent on how the handled are managed with the MSE player.

For the WebM player, it's probably not required, but I didn't want to introduce a different behaviour, so it does the same as how other decoding errors are handled.

jyavenard committed:
Errors aren't detected if media element isn't in the DOM.
https://bugs.webkit.org/show_bug.cgi?id=283150
rdar://139935149

Reviewed by Eric Carlson.

When the video element isn't in the DOM, we use a WebCoreDecompressionSession
to decode the content in place of an AVSampleBufferDisplayLayer.
In the common case with an AVSBDL, we listen to layer errors published by the NotificationCentre.
When a WebCoreDecompressionSession is in use, no such events are fired.
The WebCoreDecompressionSession provides a listener for errors, but it was unused functionality.

We add an error handler to the WebCoreDecompressionSession used by their respective MediaPlayerPrivate
which will allow the error to be handled and reported to the HTMLMediaElement.

Added tests.

* LayoutTests/media/media-source/media-source-vp8-webm-error-expected.txt: Added.
* LayoutTests/media/media-source/media-source-vp8-webm-error-offscreen-expected.txt: Added.
* LayoutTests/media/media-source/media-source-vp8-webm-error-offscreen.html: Added.
* LayoutTests/media/media-source/media-source-vp8-webm-error.html: Added.
* LayoutTests/media/media-vp8-webm-error-offscreen-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-error-offscreen.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::ensureDecompressionSession):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::ensureDecompressionSession):

Canonical link: https://commits.webkit.org/286685@main
